taropsbt/interface.go

```diff
@@ -0,0 +1,275 @@
+		return !o.IsChange
+	}
+
+	// VOutIsInteractive is a predicate that returns true if the virtual
+	// transaction is interactive.
+	VOutIsInteractive = func(o *VOutput) bool {
+		return o.Interactive
+	}
+)
+
+// VPacket is a PSBT extension packet for a virtual transaction. It represents
+// the virtual asset state transition as it will be validated by the Taro VM.
+// Some elements within the virtual packet may refer to on-chain elements (such
+// as the anchor BTC transaction that was used to anchor the input that is
+// spent). But in general a virtual transaction does NOT directly map onto a BTC
+// transaction. It is entirely possible that multiple virtual transactions will
+// be merged into a single BTC transaction. Thus, each asset state transfer is
+// represented in a virtual TX and multiple asset state transfers can be
+// anchored within a single BTC transaction.
+//
+// NOTE: A virtual transaction only carries the asset state transition for a
+// single asset ID. If multiple inputs are given, they must all belong to the
+// same asset ID (which means those asset UTXOs are being merged and possibly
+// split again in this virtual transaction). Therefore, if an anchor output
+// carries commitments for multiple assets, a virtual transaction needs to be
+// created, signed and then anchored for each asset ID separately.
+//
+// TODO(guggero): Actually support merging multiple virtual transactions into a
+// single BTC transaction.
+type VPacket struct {
+	// Inputs is the list of asset inputs that are being spent.
+	Inputs []*VInput
+
+	// Outputs is the list of new asset outputs that are created by the
+	// virtual transaction.
+	Outputs []*VOutput
+
+	// ChainParams are the Taro chain parameters that are used to encode and
+	// decode certain contents of the virtual packet.
+	ChainParams *address.ChainParams
+}
+
+// SetInputAsset sets the input asset that is being spent.
+func (p *VPacket) SetInputAsset(index int, a *asset.Asset, proof []byte) {
+	if index >= len(p.Inputs) {
+		p.Inputs = append(p.Inputs, &VInput{})
+	}
+	p.Inputs[index].asset = a.Copy()
+	p.Inputs[index].proof = proof
+}
+
+// HasSplitCommitment determines if this transaction results in an asset split.
+// This is either the case if the value of the input asset is split or if one of
+// the outputs is non-interactive (in which case we need to have a zero value
+// tombstone asset in the change output).
+func (p *VPacket) HasSplitCommitment() (bool, error) {
+	for i := 0; i < len(p.Outputs); i++ {
+		// If any of the recipient asset is nil, this virtual
+		// transaction hasn't been prepared correctly, and we cannot
+		// determine if there is a split commitment.
+		if p.Outputs[i].Asset == nil {
+			return false, fmt.Errorf("recipient asset %d is nil", i)
+		}
+
+		// If there is a split commitment witness present, we know there
+		// is a split going on.
+		if p.Outputs[i].Asset.HasSplitCommitmentWitness() {
+			return true, nil
+		}
+	}
```

Comment:
> The doc mentions inspecting input assets but the function doesn't do that. Should it?

No, this only looks at the outputs. And the doc only mentions when this is expected to be the case, not that it actually needs to look at the inputs (which wouldn't tell you this anyway, since the outputs determine whether it's a split or not).

> Should there be an equivalent for Outputs?

I don't think there is a need for it, no. The reason it was added for the input asset was to make sure we always create a copy. But because an output asset can only be created by making a copy of the input (or creating a split commitment, which internally copies the assets as well), that restriction is not necessary. The reason why an asset can be nil in the output is because when we call the `FundVirtualTransaction`, we don't yet know what the asset is supposed to look like, we only say what amount we want to send where.

> Also why not for i, output := range p.Outputs?

Good point, switched.